pkg/reconciler/base.go

```diff
@@ -273,3 +286,120 @@ func EnqueueObjectsInNamespaceOf(inf cache.SharedInformer, resyncFn filteredGlob
 		resyncFn(isInNamespace(adapter.GetNamespace()), inf)
 	}
 }
+
+// adapterPodWithAncestorOfKind returns a filter function which returns whether a Pod:
+//  - has labels that correspond to an adapter of the given type
+//  - has an outermost ancestor of the given kind
+func adapterPodWithAncestorOfKind(ctx context.Context, typ kmeta.OwnerRefable) objectFilterFunc {
+	return func(obj interface{}) bool {
+		return hasAdapterLabelsForType(typ)(obj) &&
+			hasOutermostAncestorOfKind(ctx, typ.GetGroupVersionKind())(obj)
+	}
+}
+
+// hasOutermostAncestorOfKind returns a filter function which returns whether
+// the given object has an outermost ancestor of the given kind.
+func hasOutermostAncestorOfKind(ctx context.Context, gvk schema.GroupVersionKind) objectFilterFunc {
+	return func(obj interface{}) bool {
+		object, ok := obj.(metav1.Object)
+		if !ok {
+			return false
+		}
+
+		ancestorCtlrRef := outermostAncestorControllerRef(ctx, object)
+		return ancestorCtlrRef != nil &&
+			ancestorCtlrRef.APIVersion == gvk.GroupVersion().String() &&
+			ancestorCtlrRef.Kind == gvk.Kind
+	}
+}
+
+// outermostAncestorControllerRef returns the outermost ancestor controller of
+// the given object.
+func outermostAncestorControllerRef(ctx context.Context, obj metav1.Object) *metav1.OwnerReference {
+	return resolveOutermostAncestorControllerRef(ctx, obj, schema.GroupVersionKind{})
+}
```

Comment:
`resolveOutmostAncestorControllerRef` is **recursive** (see its code right below).

However, for simplicity, because we don't want to deal with dynamic clients and prefer to look up objects in memory (through listers), the recursion is only able to look for an ancestor up that chain of dependencies: `pod` -> `replicaset` -> `deployment` -> `TriggerMesh component`.

I can't think about a case where we would need to recurse up some other object type, but if that ever becomes the case, the function can simply be extended via additional switch cases.